File: week2_algorithmic_warmup/2_last_digit_of_fibonacci_number/fibonacci_last_digit.js
// by Alexander Nikolskiy

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  terminal: false,
});

process.stdin.setEncoding("utf8");
rl.on("line", readLine);

function readLine(line) {
  console.log(fib(parseInt(line, 10)));
  process.exit();
}

function fib(n) {
  if (n <= 1) return n;

  let previous = 0;
  let current = 1;

  for (let i = 0; i < n - 1; i++) {
    let tmp_previous = previous;
    previous = current;
    current = tmp_previous + current;
  }

  return current % 10;
}
console.log(fib(15));

module.exports = fib;

File: week2_algorithmic_warmup/2_last_digit_of_fibonacci_number/FibonacciLastDigit.java
public class FibonacciLastDigit {
    private static int getFibonacciLastDigitNaive(int n) {
        if (n <= 1)
            return n;

        int previous = 0; 
        int current  = 1; 

        for (int i = 0; i < n - 1; ++i) {
            int tmp_previous = previous; 
            previous = current; 
            current = tmp_previous + current; 
        }

        return current % 10;
    }

    void findFib(int f[]) {
        f[0] = 0;
        f[1] = 1;

        for (int i = 2; i <= 59; i++) {
            f[i] = (f[i - 1] + f[i - 2]) % 10; 
        }
    }
    
    int fibLastDigit(long n) {
        int f[] = new int[60];

        findFib(f);

        int index = (int)(n % 60L);

        return f[index];
    }

    public static void main(String[] args) {
        int n = 15;
        int c = fibLastDigit(n);
        // int c = getFibonacciLastDigitNaive(n);
        System.out.println(c);
    }
}

